src/main/scala/io/iohk/ethereum/crypto/package.scala

```diff
@@ -17,6 +17,8 @@ package object crypto {
   val curveParams: X9ECParameters = SECNamedCurves.getByName("secp256k1")
   val curve: ECDomainParameters = new ECDomainParameters(curveParams.getCurve, curveParams.getG, curveParams.getN, curveParams.getH)
 
+  val kec512engine = new Keccak512
```

Comment:
Definitely looks like a speed improvement, but is it secure this way? If it is maybe we could do this for other digests as well?

Reply:
Each digest resets engine state so it should be safe in one threaded environment.  That's why i have synchronized method which uses this engine ( `kec512` is only used in miner and ethash so this should not lead to performance decrease because of thread contention)

As for `kec256` it is used in several other actors so synchronising its method could lead to thread contention  and not much performance increase. It could possibly be negated by each Actor having its own engine and passing it to the function, but i am not sure it worth it. (maybe one place i can think of which would gain from it is possibly `DiscoveryListner` as it is alocating new engine for each message decoding and encoding and some recording were pointing to that methods)